### ISSUE TITLE
BIO_connect_retry will  Endless loop in BIO_CONN_S_CONNECT  return ETIMEDOUT

### ISSUE THREAD

**Issue by yangyangtiantianlonglong** [openssl/openssl#11449]
```javascript
OpenSSL> version
OpenSSL 3.0.0-dev xx XXX xxxx
```
https://github.com/openssl/openssl/blob/f4c88073091592b1ff92ba12c894488ff7d03ece/crypto/bio/bss_conn.c#L156-L162
https://github.com/openssl/openssl/blob/f4c88073091592b1ff92ba12c894488ff7d03ece/crypto/bio/bio_lib.c#L853-L868
issue: sock connect don't stop
1.   The func conn_state  lines 158 BIO_connect return 0, get_last_sys_error return  ETIMEDOUT.
2.   The func BIO_connect_retry  lines 868 will goto retry.
3.    BIO_do_handshake --> conn_state -->BIO_do_handshake will endless loop.

**Comment by yangyangtiantianlonglong** [comment]
Why the system ETIMEDOUT does not gradually reduce the timeout time？and always endless loop connect.

**Comment by DDvO** [comment]
> Why the system ETIMEDOUT does not gradually reduce the timeout time？and always endless loop connect.

The "goto retry" is needed because of some strange behavior of `BIO_do_handshake()` at least in the blocking case.
Good point that (in the non-bocking case) the timeout condition should be checked.
I'm going to provide a fix for that.

**Comment by DDvO** [comment]
@yangyangtiantianlonglong, does the following patch solve the problem for you?
```
diff --git a/crypto/bio/bio_lib.c b/crypto/bio/bio_lib.c
index 1579f7c366..ed893e2ce2 100644
--- a/crypto/bio/bio_lib.c
+++ b/crypto/bio/bio_lib.c
@@ -867,8 +867,8 @@ int BIO_connect_retry(BIO *bio, int timeout)
              * unless using BIO_reset(), blocking next connect() may crash and
              * non-blocking next BIO_do_handshake() will fail
              */
-            goto retry;
-        } else if (BIO_should_retry(bio)) {
+        }
+        if (get_last_sys_error() == ETIMEDOUT || BIO_should_retry(bio)) {
             /* will not actually wait if timeout == 0 (i.e., blocking BIO) */
             rv = bio_wait(bio, max_time, 100 /* milliseconds */);
             if (rv > 0)

```

**Comment by DDvO** [comment]
This issue should be fixed by #11986.
